File: src/ProgramableNetwork/ModuleParser/Registrator/ControllerTemplate.cs
using Mafi;
using System.Collections.Generic;

namespace ProgramableNetwork.Python
{
    public struct ControllerTemplate
    {
        public delegate void Settings();
        public delegate Settings ModulePlacement(Controller controller);

        public string id;
        public string name;
        public string description;
        public ColorRgba color;
        public ModulePlacement modules;

        public ControllerTemplate(string id, string name, string description, ColorRgba color, ModulePlacement modules)
        {
            this.id = id;
            this.name = name;
            this.description = description;
            this.color = color;
            this.modules = modules;
        }

        public override bool Equals(object obj)
        {
            return obj is ControllerTemplate other &&
                   id == other.id &&
                   name == other.name &&
                   description == other.description &&
                   EqualityComparer<ModulePlacement>.Default.Equals(modules, other.modules);
        }

        public override int GetHashCode()
        {
            int hashCode = -1055989087;
            hashCode = hashCode * -1521134295 + EqualityComparer<string>.Default.GetHashCode(id);
            hashCode = hashCode * -1521134295 + EqualityComparer<string>.Default.GetHashCode(name);
            hashCode = hashCode * -1521134295 + EqualityComparer<string>.Default.GetHashCode(description);
            hashCode = hashCode * -1521134295 + EqualityComparer<ModulePlacement>.Default.GetHashCode(modules);
            return hashCode;
        }

        public void Deconstruct(out string id, out string name, out string description, out ColorRgba color, out ModulePlacement modules)
        {
            id = this.id;
            name = this.name;
            description = this.description;
            color = this.color;
            modules = this.modules;
        }
    }
}

File: src/ProgramableNetwork.Modules/Custom/controller_template.py

from Core.template import Controller
from Core.mafi import fix

# How to get IDS, see Core.ids
# Even is not Id defined inside Core.ids, will be correctly displayed

# used modules
from Core.ids import Arithmetic_Display_7SEG_B, Connection_Display_7SEG_B, Runtime_Clock_2
from Mafi import ColorRgba

# tier contains 16 columns, 4 rows
tier = 1

# Note image must be located within:
# Assets/ProgramableNetwork/Computer/Icon_{id}.png inside modification directory or asset bundle
# in this example: Assets/ProgramableNetwork/Computer/Icon_ClockExample.png
class ClockExample(Controller):
    name = "Basic hour clock"
    description = "Creates an controller configuration for recording time from point in which was controller built."
    color = ColorRgba.CornflowerBlue
    
    def modules(self):
        clock = self.add_module(Runtime_Clock_2, 0, 0);

        converterA = self.add_module(Arithmetic_Display_7SEG_B, 2, 0)
        converterB = self.add_module(Arithmetic_Display_7SEG_B, self.next_row, self.next_column)
        converterC = self.add_module(Arithmetic_Display_7SEG_B, self.next_row, self.next_column)
        converterD = self.add_module(Arithmetic_Display_7SEG_B, self.next_row, self.next_column)
        converterE = self.add_module(Arithmetic_Display_7SEG_B, self.next_row, self.next_column)
        converterF = self.add_module(Arithmetic_Display_7SEG_B, self.next_row, self.next_column)

        displayA = self.add_module(Connection_Display_7SEG_B, 3, 0)
        displayB = self.add_module(Connection_Display_7SEG_B, self.next_row, self.next_column)
        displayC = self.add_module(Connection_Display_7SEG_B, self.next_row, self.next_column)
        displayD = self.add_module(Connection_Display_7SEG_B, self.next_row, self.next_column)
        displayE = self.add_module(Connection_Display_7SEG_B, self.next_row, self.next_column)
        displayF = self.add_module(Connection_Display_7SEG_B, self.next_row, self.next_column)

        displayA["N"] = converterA["bits"]
        displayB["N"] = converterB["bits"]
        displayC["N"] = converterC["bits"]
        displayD["N"] = converterD["bits"]
        displayE["N"] = converterE["bits"]
        displayF["N"] = converterF["bits"]

        converterE["V"] = converterF["rest"]
        converterD["V"] = converterE["rest"]
        converterC["V"] = converterD["rest"]
        converterB["V"] = converterC["rest"]
        converterA["V"] = converterB["rest"]

        converterF["V"] = clock["clock"]

        # forward modules for settings phase
        return [clock]

    def settings(self, modules):
        modules[0].Field.set_int("max_count", 24*60*60)
        modules[0].Field.set_int("clock_period", 10) # each game second